bin/update_repo.sh

```diff
@@ -25,6 +25,17 @@ case $TAG in
           * ) usage ;;
 esac
 
+# Prevent simultaneous mash runs from colliding
+# Causes errors when another instance opens an incompletely downloaded RPM
+# Wait up to 5 minutes for the other task to complete
+# http://mywiki.wooledge.org/BashFAQ/045
+lockfile="/tmp/lock.update_repo-$SERIES.$DVER"
+exec 99>$lockfile
+if ! flock --wait 300 99  ; then
+         printf 'another instance is running\n';
+         exit 1
+fi
+
```

Comment:
Eh, so, update_all_repos.sh already does locking, and that is the entry point for running this script (update_repo.sh).

https://github.com/opensciencegrid/osg-repo-scripts/blob/v1.9.1/bin/update_all_repos.sh#L38-L42

So I kinda feel like this new section can be dropped altogether.

```suggestion
```

...

But, in a hypothetical scenario where for some reason (??) you had a desire to run update_repo.sh manually to update an individual repo without potentially interfering with a different run, i would have suggested something like the following:

```suggestion
# Prevent simultaneous mash runs from colliding
lockfile=/var/lock/repo/update_repo-$SERIES.$DVER.$REPO
exec 299> "$lockfile"
if ! flock -n 299; then
  echo "Can't acquire lock, is $(basename "$0") already running?" >&2
  exit 1
fi

```

- put the lockfile under /var/lock/repo, _not_ under /tmp
- include `$REPO` in the lockfile name
- use echo (rather than printf) to match the rest of the script
- output failure message to stderr
- do _not_ wait five minutes.  If another instance is running, it's better to be informed of that immediately.  And if this was a scheduled instance and there is a previous run still going, let the first one complete and have that count as the completion for this run.  I don't think there's any need to queue them up.

However, just to be clear, i am _not_ suggesting you apply the second set of changes above.  I am suggesting you just drop this lock section since it is handled by the calling script.

Reply:
I'm indeed trying for the hypothetical scenario! The goal was to quickly complete multiple mash runs, getting a k8s instance of repo.osg from blank to ready. (...in 30 minutes or less, or your download is free.) Mash itself seems limited by single-CPU performance, and running simultaneous copies gives a decent speed-up.

In the k8s initContainer phase, GNU parallel can simultaneously run `update_repo.sh` on the osg-tags list:
```
parallel osg-tags --retries 10 update_repo.sh {}
```

The 5-minute lock wait goes with the `--retries 10`. If locking fails, parallel will move on to other tags and retry later.

The lockfile name intentionally leaves out `$REPO`.  Mash throws errors if simultaneous runs modify the same cached files (under `/usr/local/mash/cache`). Locking on "$SERIES.$DVER" (e.g. `3.5-el7`) rather than "$SERIES.$DVER.$REPO" (e.g. `3.5-el7-testing`) was necessary, because the same package could be in multiple repos.

I'm totally up for refactoring these changes into something less insane.

Reply:
Thanks for explaining!!

I didn't realize the parallel stuff was going on.  John explained it's happening in the container:

https://github.com/opensciencegrid/tiger-osg-config/blob/ee0437c/manifests/base/osg-repo/deployment.yaml#L37

If this is working for you now, that's fine for now.

As i am thinking about it, i feel like a script that does not require per series-dver locking & timeouts should be possible.  Maybe when management is not looking i'll take a whack at a burst-mode start up script that does ... whatever is allowed in parallel.  And then that script could simply grab the all-repos lock (as in [update_all_repos.sh](https://github.com/opensciencegrid/osg-repo-scripts/blob/v1.9.1/bin/update_all_repos.sh#L38-L42)).

The existence of timeouts suggests to me there is a better way to do it.  But we don't have to find that today.

Another (hopefully minor) detail is that technically i think the same rpm can be tagged into more than one series.  I think it's rare except for when we are bootstrapping for builds for the next osg series.  I haven't thought about it long enough but there might be a scenario where this breaks due to rpms that appear in multiple series.  For now maybe it's enough just to be aware of it though.

...

Otherwise, if you want to keep your current locking setup, i think i would still ask that you at least make the small tweaks for the lock dir and the error message output:

```suggestion
# Prevent simultaneous mash runs from colliding
# Causes errors when another instance opens an incompletely downloaded RPM
# Wait up to 5 minutes for the other task to complete
lockfile=/var/lock/repo/lock.update_repo-$SERIES.$DVER
exec 99>$lockfile
if ! flock --wait 300 99  ; then
         echo "another instance is running" >&2
         exit 1
fi

```

Reply:
I think we need to create the directory too?
https://github.com/opensciencegrid/osg-repo-scripts/pull/19/commits/d9b95ef66e1e398f26581abcc70375b5398310eb

Reply:
Ah ok.  Yeah before this would be created in the calling script (update_all_repos.sh) if it didn't exist, but i guess now you are calling it separately.